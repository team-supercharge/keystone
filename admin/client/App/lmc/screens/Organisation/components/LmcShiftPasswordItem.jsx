import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Button, GlyphButton } from '../../../../elemental'
import { Link } from 'react-router'
import ConfirmationDialog from '../../../../shared/ConfirmationDialog'

export default class LmcShiftPasswordItem extends Component {
    state = {
        deleteDialogOpen: false
    }

    toggleDeleteDialog = () => {
        this.setState(prevProps => ({
            deleteDialogOpen: !prevProps.deleteDialogOpen
        }))
    }

    handleDeleteConfirm = (id) => {
        this.toggleDeleteDialog()
        this.props.onDelete(id)
    }

    render () {
        const { shift } = this.props
        return (
            <li style={styles.container}>
                <div style={styles.entry}>
                    <span>
                        { shift.title }
                    </span>
                </div>
                <div style={styles.entry}>
                    <span>
                        { shift.startTime } - { shift.endTime }
                    </span>
                </div>
                <div style={styles.entry}>
                    <span>
                        { shift.length } hours
                    </span>
                </div>
                <div style={styles.entryActions}>
                    <GlyphButton
                        component={Link}
                        glyph='pencil'
                        position='left'
                        to={`${Keystone.adminPath}/shifts/${shift.id}`}
                        style={styles.button}
                    >
                        Edit
                    </GlyphButton>
                    <Button 
                        onClick={this.toggleDeleteDialog} 
                        variant="link" 
                        color="delete"
                        style={styles.button}
                        data-button="delete"
                    >
                        Delete
                    </Button>
                    <ConfirmationDialog
                        confirmationLabel='Delete'
                        confirmationType='warning'
                        isOpen={this.state.deleteDialogOpen}
                        onCancel={this.toggleDeleteDialog}
                        onConfirmation={() => this.handleDeleteConfirm(shift.id)}
                    >
                        <div>
                            Are you sure you want to delete this shift? 
                            If you go ahead, it canâ€™t be undone.
                        </div>
                    </ConfirmationDialog>
                </div>
            </li>
        ) 
    }
}

const styles = {
    container: {
        width: '100%',
        minWidth: 290,
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginTop: 12,
    },
    button: {
        width: 76,
    },
    entry: {
        width: '25%',
        minWidth: 70,
    },
    entryActions: {
        width: '25%',
        minWidth: 70,
        position: 'relative',
        bottom: 5,
    }
}

LmcShiftPasswordItem.propTypes = {
    shift: PropTypes.object,
    onDelete: PropTypes.func.isRequired
}