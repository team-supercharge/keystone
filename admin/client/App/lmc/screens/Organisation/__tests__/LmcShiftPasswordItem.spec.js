import React from 'react'
import { shallow } from 'enzyme'
import LmcShiftPasswordItem from '../components/LmcShiftPasswordItem.jsx'

describe('LmcShiftPasswordItem', () => {
    let wrapper
    let shift
    let savedKeystone
    const onDeleteMock = jest.fn()

    beforeAll(() => {
        savedKeystone = global.Keystone
        global.Keystone = { adminPath: '/admin' }
    })

    beforeEach(() => {
        shift = {
            id: 'testId',
            title: 'testShift',
            password: 'testPassword',
            startTime: '08:00',
            endTime: '18:00',
            accessTime: '4'
        }
        wrapper = shallow(
            <LmcShiftPasswordItem
                shift={shift}
                onDelete={onDeleteMock}
            />
        )
    })

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('renders a list item', () => {
        expect(wrapper.find('li').length).toEqual(1)
    })

    it('renders a confirmation dialog to trigger its onDelete prop', () => {
        const dialog = wrapper.find('ConfirmationDialog')
        dialog.props().onConfirmation()
        expect(onDeleteMock).toHaveBeenCalledTimes(1)
        expect(onDeleteMock).toHaveBeenCalledWith(shift.id)
    })

    it('renders a button to edit the shift', () => {
        const button = wrapper.find('GlyphButton')
        expect(button.props().to).toEqual(`${Keystone.adminPath}/shifts/${shift.id}`)
    })

    afterEach(() => {
        onDeleteMock.mockClear()
    })

    afterAll(() => {
        global.Keystone = savedKeystone
    })
})