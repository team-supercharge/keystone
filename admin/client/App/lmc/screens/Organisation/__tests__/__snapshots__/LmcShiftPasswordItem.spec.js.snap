// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcShiftPasswordItem renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 12,
        "minWidth": 290,
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "minWidth": 70,
          "width": "25%",
        }
      }
    >
      <span>
        testShift
      </span>
    </div>
    <div
      style={
        Object {
          "minWidth": 70,
          "width": "25%",
        }
      }
    >
      <span>
        08:00
         - 
        18:00
      </span>
    </div>
    <div
      style={
        Object {
          "minWidth": 70,
          "width": "25%",
        }
      }
    >
      <span>
         hours
      </span>
    </div>
    <div
      style={
        Object {
          "bottom": 5,
          "minWidth": 70,
          "position": "relative",
          "width": "25%",
        }
      }
    >
      <GlyphButton
        component={[Function]}
        glyph="pencil"
        glyphStyle={Object {}}
        position="left"
        style={
          Object {
            "width": 76,
          }
        }
        to="/admin/shifts/testId"
      >
        Edit
      </GlyphButton>
      <Button
        aphroditeStyles={Array []}
        color="delete"
        data-button="delete"
        onClick={[Function]}
        style={
          Object {
            "width": 76,
          }
        }
        variant="link"
      >
        Delete
      </Button>
      <ConfirmationDialog
        cancelLabel="Cancel"
        confirmationLabel="Delete"
        confirmationType="warning"
        isOpen={false}
        onCancel={[Function]}
        onConfirmation={[Function]}
      >
        <div>
          Are you sure you want to delete this shift? If you go ahead, it can’t be undone.
        </div>
      </ConfirmationDialog>
    </div>
  </li>,
  "nodes": Array [
    <li
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 12,
          "minWidth": 290,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "minWidth": 70,
            "width": "25%",
          }
        }
      >
        <span>
          testShift
        </span>
      </div>
      <div
        style={
          Object {
            "minWidth": 70,
            "width": "25%",
          }
        }
      >
        <span>
          08:00
           - 
          18:00
        </span>
      </div>
      <div
        style={
          Object {
            "minWidth": 70,
            "width": "25%",
          }
        }
      >
        <span>
           hours
        </span>
      </div>
      <div
        style={
          Object {
            "bottom": 5,
            "minWidth": 70,
            "position": "relative",
            "width": "25%",
          }
        }
      >
        <GlyphButton
          component={[Function]}
          glyph="pencil"
          glyphStyle={Object {}}
          position="left"
          style={
            Object {
              "width": 76,
            }
          }
          to="/admin/shifts/testId"
        >
          Edit
        </GlyphButton>
        <Button
          aphroditeStyles={Array []}
          color="delete"
          data-button="delete"
          onClick={[Function]}
          style={
            Object {
              "width": 76,
            }
          }
          variant="link"
        >
          Delete
        </Button>
        <ConfirmationDialog
          cancelLabel="Cancel"
          confirmationLabel="Delete"
          confirmationType="warning"
          isOpen={false}
          onCancel={[Function]}
          onConfirmation={[Function]}
        >
          <div>
            Are you sure you want to delete this shift? If you go ahead, it can’t be undone.
          </div>
        </ConfirmationDialog>
      </div>
    </li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcShiftPasswordItem
        onDelete={[MockFunction]}
        shift={
          Object {
            "accessTime": "4",
            "endTime": "18:00",
            "id": "testId",
            "password": "testPassword",
            "startTime": "08:00",
            "title": "testShift",
          }
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcShiftPasswordItem {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleDeleteConfirm": [Function],
        "props": Object {
          "onDelete": [MockFunction],
          "shift": Object {
            "accessTime": "4",
            "endTime": "18:00",
            "id": "testId",
            "password": "testPassword",
            "startTime": "08:00",
            "title": "testShift",
          },
        },
        "refs": Object {},
        "state": Object {
          "deleteDialogOpen": false,
        },
        "toggleDeleteDialog": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <li
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginTop": 12,
              "minWidth": 290,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "minWidth": 70,
                "width": "25%",
              }
            }
          >
            <span>
              testShift
            </span>
          </div>
          <div
            style={
              Object {
                "minWidth": 70,
                "width": "25%",
              }
            }
          >
            <span>
              08:00
               - 
              18:00
            </span>
          </div>
          <div
            style={
              Object {
                "minWidth": 70,
                "width": "25%",
              }
            }
          >
            <span>
               hours
            </span>
          </div>
          <div
            style={
              Object {
                "bottom": 5,
                "minWidth": 70,
                "position": "relative",
                "width": "25%",
              }
            }
          >
            <GlyphButton
              component={[Function]}
              glyph="pencil"
              glyphStyle={Object {}}
              position="left"
              style={
                Object {
                  "width": 76,
                }
              }
              to="/admin/shifts/testId"
            >
              Edit
            </GlyphButton>
            <Button
              aphroditeStyles={Array []}
              color="delete"
              data-button="delete"
              onClick={[Function]}
              style={
                Object {
                  "width": 76,
                }
              }
              variant="link"
            >
              Delete
            </Button>
            <ConfirmationDialog
              cancelLabel="Cancel"
              confirmationLabel="Delete"
              confirmationType="warning"
              isOpen={false}
              onCancel={[Function]}
              onConfirmation={[Function]}
            >
              <div>
                Are you sure you want to delete this shift? If you go ahead, it can’t be undone.
              </div>
            </ConfirmationDialog>
          </div>
        </li>,
        "_debugID": 2,
        "_renderedOutput": <li
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginTop": 12,
              "minWidth": 290,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "minWidth": 70,
                "width": "25%",
              }
            }
          >
            <span>
              testShift
            </span>
          </div>
          <div
            style={
              Object {
                "minWidth": 70,
                "width": "25%",
              }
            }
          >
            <span>
              08:00
               - 
              18:00
            </span>
          </div>
          <div
            style={
              Object {
                "minWidth": 70,
                "width": "25%",
              }
            }
          >
            <span>
               hours
            </span>
          </div>
          <div
            style={
              Object {
                "bottom": 5,
                "minWidth": 70,
                "position": "relative",
                "width": "25%",
              }
            }
          >
            <GlyphButton
              component={[Function]}
              glyph="pencil"
              glyphStyle={Object {}}
              position="left"
              style={
                Object {
                  "width": 76,
                }
              }
              to="/admin/shifts/testId"
            >
              Edit
            </GlyphButton>
            <Button
              aphroditeStyles={Array []}
              color="delete"
              data-button="delete"
              onClick={[Function]}
              style={
                Object {
                  "width": 76,
                }
              }
              variant="link"
            >
              Delete
            </Button>
            <ConfirmationDialog
              cancelLabel="Cancel"
              confirmationLabel="Delete"
              confirmationType="warning"
              isOpen={false}
              onCancel={[Function]}
              onConfirmation={[Function]}
            >
              <div>
                Are you sure you want to delete this shift? If you go ahead, it can’t be undone.
              </div>
            </ConfirmationDialog>
          </div>
        </li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcShiftPasswordItem
    onDelete={[MockFunction]}
    shift={
      Object {
        "accessTime": "4",
        "endTime": "18:00",
        "id": "testId",
        "password": "testPassword",
        "startTime": "08:00",
        "title": "testShift",
      }
    }
  />,
}
`;
